Meta: Update CMake for specifing jakt files to build

robryanx opened the pull request.

ADKaster commented on Userland/Utilities/CMakeLists.txt:
```diff
@@ -52,16 +52,20 @@ foreach(CMD_SRC ${CMD_SOURCES})
-        TARGETS hello-jakt
-    )
+    file(GLOB CMD_SOURCES_JAKT  CONFIGURE_DEPENDS "*.jakt")
+    foreach(CMD_SRC_JAKT ${CMD_SOURCES_JAKT})
```
`foreach(CMD_SRC_JAKT IN LISTS CMD_SRCS_JAK)`

robryanx commented on Userland/Utilities/CMakeLists.txt:
```diff
@@ -52,16 +52,20 @@ foreach(CMD_SRC ${CMD_SOURCES})
-        RECOMMENDED
-        TARGETS hello-jakt
-    )
+    file(GLOB CMD_SOURCES_JAKT  CONFIGURE_DEPENDS "*.jakt")
```
Yeah that make sense, I have changed this pull request to provide functions for manually declaring files/ directories.

ADKaster commented on Meta/CMake/utils.cmake:
```diff
@@ -221,3 +221,33 @@ function(download_file url path)
+    )
+endfunction()
+
+function(add_directory_jakt directory)
```
This looks much more reasonable. Though, this could be simplified further into one command that takes the jakt file and the executable name, instead of two commands with a bunch of copy paste.

e.g.
`function(add_jakt_executable exe_name source_file)`

robryanx replied:
Yeah, here is a 3rd version like that.
Also added the idea of a YAKT_TARGETS list as a way of dealing with having a c++ and yakt version of something.

ADKaster replied:
Hmm. I'm not convinced we should add the JAKT_TARGETS workaround until we actually have some JAKT_TARGETS to add. Or at least, if we do keep it here in this PR, a comment explaining that this is the intended use should be used.

ADKaster replied:
oh I missed that you added hello-jakt to the list.

Ok, just a comment over it then :D then I think this is good to go. Thanks for adding a better helper!

robryanx replied:
Sure, have added that.

AtkinsSJ commented:
Typo in the second commit message and code: "YAKT_TARGETS" should be "JAKT_TARGETS".

robryanx commented:
> Typo in the second commit message and code: "YAKT_TARGETS" should be "JAKT_TARGETS".

Lol yes, fixed now.

drunderscore commented:
Sorry to pile on another typo 😅, but:
Typo in second commit message: `Utilites` -> `Utilities`

linusg commented on Userland/Utilities/CMakeLists.txt:
```diff
@@ -11,59 +11,54 @@ list(APPEND RECOMMENDED_TARGETS
-# FIXME: Support specifying component dependencies for utilities (e.g. WebSocket for telws)
+set(JAKT_TARGETS "")
+if (ENABLE_JAKT)
+    add_jakt_executable(hello-jakt hello-jakt.jakt)
```
Why the rename?

robryanx replied:
No reason now, previously I was looking at having it just use the filename for the executable name for a single file, changed it back.

robryanx commented:
> Sorry to pile on another typo sweat_smile, but: Typo in second commit message: `Utilites` -> `Utilities`

Thanks fixed that

ADKaster reviewed (approved):
Looks good! can you resolve the conversations that are resolved/outdated so that a maintainer can see that there's no open conversations? the author is the only person who can do that on GitHub.

robryanx commented:
Done.

FireFox317 commented:
Something has gone wrong regarding the commit messages of the 2 commits. Both commit messages are exactly the same, but obviously they are different commits, and thus should have different commit messages :)

robryanx commented:
> Something has gone wrong regarding the commit messages of the 2 commits. Both commit messages are exactly the same, but obviously they are different commits, and thus should have different commit messages :)

Yeah thanks for pointing that out, got into a bit of a mess with the changes, fixed now.

linusg commented on Meta/CMake/utils.cmake:
```diff
@@ -221,3 +221,18 @@ function(download_file url path)
+    target_link_libraries(${exe_name} LibC)
+    set_target_properties(${exe_name} PROPERTIES EXCLUDE_FROM_ALL TRUE)
+    install(TARGETS ${exe_name} RUNTIME DESTINATION bin OPTIONAL)
+    serenity_component(
+        ${exe_name}
+        RECOMMENDED
+        TARGETS ${exe_name}
+    )
```
I don't think this should be part of this function, not all will be `RECOMMENDED`

robryanx replied:
I have updated `add_jakt_executable` to have an optional 3rd argument to specify `RECOMMENDED`/ `REQUIRED`

I think the CI failure on this push was an intermittent issue with package downloads, not sure how I would rerun?

robryanx commented:
I added a couple of changes to this:
- `add_jakt_executable` will assume a full path instead of the current source directory for `source_file` if the filename starts with a forward slash
- Added additional required compiler options in `compile_jakt`

When combined with https://github.com/SerenityOS/jakt/pull/969 this now allows for example the selfhost compiler to be built:
`add_jakt_executable(jakt ${JAKT_SOURCE_DIR}/selfhost/main.jakt)`

robryanx commented:
Going to close this and make a new pull request as it makes sense now to extend this to account for the new bootstrap cpp in place of the rust compiler.
